crates/api_crud/src/user/create.rs

```diff
@@ -238,3 +227,319 @@ pub async fn register(
+
+  let local_user: LocalUser;
+  if let Some(user_view) = local_user_view {
+    // user found by oauth_user_id => Login user
+    local_user = user_view.clone().local_user;
+
+    check_user_valid(&user_view.person)?;
+    check_email_verified(&user_view, &site_view)?;
+    check_registration_application(&user_view, &site_view.local_site, &mut context.pool()).await?;
+  } else {
+    // user has never previously registered using oauth
+
+    // prevent registration if registration is closed
+    if local_site.registration_mode == RegistrationMode::Closed {
+      Err(LemmyErrorType::RegistrationClosed)?
+    }
+
+    // prevent registration if registration is closed for OAUTH providers
+    if !local_site.oauth_registration {
+      return Err(LemmyErrorType::OauthRegistrationClosed)?;
+    }
+
+    // Extract the OAUTH email claim from the returned user_info
+    let email: String;
+    if let Some(email_value) = user_info.get("email") {
+      email = serde_json::from_value::<String>(email_value.clone())
+        .ok()
+        .ok_or(LemmyErrorType::OauthLoginFailed)?;
+    } else {
+      return Err(LemmyErrorType::OauthLoginFailed)?;
+    }
+
+    // Lookup user by OAUTH email and link accounts
+    local_user_view = LocalUserView::find_by_email(&mut context.pool(), &email).await?;
+
+    let person;
+    if let Some(user_view) = local_user_view {
+      // user found by email => link and login if linking is allowed
+
+      if oauth_provider.account_linking_enabled {
+        // Link with OAUTH => Login user
+        let oauth_account_form =
+          OAuthAccountInsertForm::new(user_view.local_user.id, oauth_provider.id, oauth_user_id);
+
+        OAuthAccount::create(&mut context.pool(), &oauth_account_form)
+          .await
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+
+        local_user = user_view.local_user.clone();
+
+        check_user_valid(&user_view.person)?;
+        check_registration_application(&user_view, &site_view.local_site, &mut context.pool())
+          .await?;
+        check_email_verified(&user_view, &site_view)?;
+      } else {
+        // email already registered
+        return Err(LemmyErrorType::EmailAlreadyExists)?;
+      }
+    } else {
+      // No user was found by email => Register as new user
+
+      // Extract the OAUTH name claim from the returned user_info
+      let user_name: String;
+      if let Some(user_name_value) = user_info.get(oauth_provider.name_claim) {
+        user_name = serde_json::from_value::<String>(user_name_value.clone())
+          .ok()
+          .ok_or(LemmyErrorType::OauthLoginFailed)?;
+      } else {
+        return Err(LemmyErrorType::OauthLoginFailed)?;
+      }
+
+      // remove spaces from username
+      let username = str::replace(&user_name, " ", "_");
+      let username = username + &rand::thread_rng().gen_range(0..999).to_string();
```

Comment:
Username with random number attached will look a bit ugly. I would only do this if the original username is already taken. And even then its theoretically possible that the generated username is already taken, so you may have to loop until a free username is found.

Reply:
Actually I would just take the name from `data.username` instead of oauth.

Reply:
I assume it can be passed in the same way as the registration application.

Reply:
If the OAuth AS doesn't provide a username, does the user get prompted to enter one?

(I'm possibly looking at the SSO signup path for Mastodon)